fairscale/nn/data_parallel/fully_sharded_data_parallel.py

```diff
@@ -1381,7 +1381,11 @@ def _queue_wait_for_post_backward(self) -> None:
     def _wait_for_post_backward(self) -> None:
         """Wait for post-backward to finish. Only called on root instance."""
         assert self._is_root
-        if self._has_params:
+        # Check if the root module has params and if any of them has
+        # the `requires_grad` field set. If `requires_grad=False` for
+        # all the params, the post_backward hook will not fire and the
+        # state will remain in `TrainingState.BACKWARD_PRE`.
+        if self._has_params and any([p.requires_grad for p in self.params]):
```

Comment:
perhaps self._has_params is not needed since in that case self.params will just be an empty list?